tests/cli/test_scratch.py

```diff
@@ -0,0 +1,215 @@
+    file_path = directory_path / str(uuid.uuid4())
+    with file_path.open("w") as stream:
+        stream.write("foo")
+    yield file_path
+    os.remove(file_path)
+
+
+@pytest.fixture
+def nonexistant_path(directory_path: Path) -> Path:
+    file_path = directory_path / str(uuid.uuid4())
+    assert not file_path.exists()
+    return file_path
+
+
+@patch("blueapi.cli.scratch.Popen")
+def test_scratch_install_installs_path(
+    mock_popen: Mock,
+    directory_path: Path,
+):
+    mock_process = Mock()
+    mock_process.returncode = 0
+    mock_popen.return_value = mock_process
+
+    scratch_install(directory_path, timeout=1.0)
+
+    mock_popen.assert_called_once_with(
+        [
+            "python",
+            "-m",
+            "pip",
+            "install",
+            "--no-deps",
+            "-e",
+            str(directory_path),
+        ]
+    )
+
+
+def test_scratch_install_fails_on_file(file_path: Path):
+    with pytest.raises(KeyError):
+        scratch_install(file_path, timeout=1.0)
+
+
+def test_scratch_install_fails_on_nonexistant_path(nonexistant_path: Path):
+    with pytest.raises(KeyError):
+        scratch_install(nonexistant_path, timeout=1.0)
+
+
+@patch("blueapi.cli.scratch.Popen")
+@pytest.mark.parametrize("code", [1, 2, 65536])
+def test_scratch_install_fails_on_non_zero_exit_code(
+    mock_popen: Mock,
+    directory_path: Path,
+    code: int,
+):
+    mock_process = Mock()
+    mock_process.returncode = code
+    mock_popen.return_value = mock_process
+
+    with pytest.raises(RuntimeError):
+        scratch_install(directory_path, timeout=1.0)
+
+
+@patch("blueapi.cli.scratch.Repo")
+def test_repo_not_cloned_and_validated_if_found_locally(
+    mock_repo: Mock,
+    directory_path: Path,
+):
+    ensure_repo("http://example.com/foo.git", directory_path)
+    mock_repo.assert_called_once_with(directory_path)
+    mock_repo.clone_from.assert_not_called()
+
+
+@patch("blueapi.cli.scratch.Repo")
+def test_repo_cloned_if_not_found_locally(
+    mock_repo: Mock,
+    nonexistant_path: Path,
+):
+    ensure_repo("http://example.com/foo.git", nonexistant_path)
+    mock_repo.assert_not_called()
+    mock_repo.clone_from.assert_called_once_with(
+        "http://example.com/foo.git", nonexistant_path
+    )
+
+
+@patch("blueapi.cli.scratch.Repo")
+def test_repo_cloned_with_correct_umask(
+    mock_repo: Mock,
+    directory_path: Path,
+):
+    repo_root = directory_path / "foo"
+    file_path = repo_root / "a"
+
+    def write_repo_files():
+        repo_root.mkdir()
+        with file_path.open("w") as stream:
+            stream.write("foo")
+
+    mock_repo.clone_from.side_effect = lambda url, path: write_repo_files()
+
+    ensure_repo("http://example.com/foo.git", repo_root)
+    assert file_path.exists()
+    assert file_path.is_file()
+    st = os.stat(file_path)
+    # For some reason the mode has two extra bytes at the front
+    assert (st.st_mode & 0x00FF) & stat.S_IWGRP
```

Comment:
```suggestion
    assert st.st_mode & stat.S_IWGRP
```
Think the higher octets can be ignored.